quic/s2n-quic-transport/src/space/datagram.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    endpoint,
+    stream::{AbstractStreamManager, StreamTrait as Stream},
+    transmission::WriteContext,
+};
+use s2n_quic_core::{
+    datagram::{Endpoint, Sender},
+    frame,
+};
+
+// Contains the datagram sender and receiver implementations.
+//
+// Used to call datagram callbacks during packet transmission and
+// packet processing.
+pub struct Manager<Config: endpoint::Config> {
+    sender: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Sender,
+    // TODO: Remove this warning once Receiver is implemented
+    #[allow(dead_code)]
+    receiver: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Receiver,
+}
+
+impl<Config: endpoint::Config> Manager<Config> {
+    pub fn new(
+        sender: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Sender,
+        receiver: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Receiver,
+    ) -> Self {
+        Self { sender, receiver }
+    }
+
+    /// A callback that allows users to write datagrams directly to the packet.
+    pub fn on_transmit<S: Stream, W: WriteContext>(
+        &mut self,
+        context: &mut W,
+        stream_manager: &mut AbstractStreamManager<S>,
+    ) {
+        let mut packet = Packet {
+            context,
+            pending_streams: stream_manager.pending_streams(),
+        };
+        self.sender.on_transmit(&mut packet);
+    }
+}
+struct Packet<'a, C: WriteContext> {
+    context: &'a mut C,
+    pending_streams: bool,
+}
+
+const FRAME_TYPE_LEN: usize = 1;
+
+impl<'a, C: WriteContext> s2n_quic_core::datagram::Packet for Packet<'a, C> {
+    /// Returns the remaining space in the packet
+    fn remaining_capacity(&self) -> usize {
+        let space = self.context.remaining_capacity();
+        //= https://www.rfc-editor.org/rfc/rfc9221#section-4
+        //# The least significant bit of the Type field in the DATAGRAM frame is
+        //# the LEN bit (0x01), which indicates whether there is a Length field
+        //# present: if this bit is set to 0, the Length field is absent and the
+        //# Datagram Data field extends to the end of the packet; if this bit is
+        //# set to 1, the Length field is present.
+        // In the case where the user writes the largest datagram possible
+        // we don't factor in the size of the Length field as it will be the last
+        // frame in the packet. Thus we only have to factor out the frame type field.
+        space - FRAME_TYPE_LEN
```

Comment:
I think this needs to include the maximum length value as part of the length trimmed off. Otherwise, we're going to go write something that won't actually fit.

Also we need to saturate in case there isn't any space left.

```suggestion
        space.saturating_len(FRAME_TYPE_LEN)
```

Reply:
What is saturating space? I don't see that term anywhere in our library.